src/actions/search.js

```diff
@@ -0,0 +1,51 @@
+// @flow
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Redux actions for the search state
+ * @module actions/search
+ */
+
+import { searchSource } from "../utils/search/project-search";
+
+import { getSources, getSearchResult } from "../selectors";
+
+import type { ThunkArgs } from "./types";
+
+export function addSearchQuery(query) {
+  return ({ dispatch, getState }: ThunkArgs) => {
+    dispatch({ type: "ADD_QUERY", query });
+  };
+}
+
+export function removeSearchQuery() {
+  return ({ dispatch, getState }: ThunkArgs) => {
+    dispatch({ type: "REMOVE_QUERY" });
+  };
+}
+
+export function searchSources(query) {
+  return async ({ dispatch, getState }: ThunkArgs) => {
+    dispatch(addSearchQuery(query));
+
+    const sources = getSources(getState());
+    const validSources = sources
+      .valueSeq()
+      .filter(source => source.has("text"))
+      .toJS();
+
+    validSources.forEach(source => {
```

Comment:
perhaps:

```js
for (const source of validSources) {
  const matches = await searchSource(source, query);
  dispatch({ ya da ya da });
}
```

This strategy will let us defer the work to the worker. each search will be a new worker call and will wait for the response to come back...